### PR TITLE
Add clear app data before each test option

### PR DESCRIPTION
**Pull request by rhomel**
I added a `--clear-app-data` CLI option that will run the `pm clear` command before each test run. This can be used to ensure each instrumentation test starts the app under test in a clean initial state.

### CODE REVIEW

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import com.android.ddmlib.testrunner.RemoteAndroidTestRunner;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Run the "pm clear" command before every `run` invocation.
+ */
+public class SpoonClearingAndroidTestRunner extends RemoteAndroidTestRunner {
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java]: Let's just rename this `SpoonAndroidTestRunner`

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import com.android.ddmlib.TimeoutException;
+import com.android.ddmlib.testrunner.ITestRunListener;
+import com.android.ddmlib.testrunner.RemoteAndroidTestRunner;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Run the "pm clear" command before every `run` invocation.
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java]: `Allows the option of running the "pm clear" command before every "run" invocation`

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ */
+public class SpoonClearingAndroidTestRunner extends RemoteAndroidTestRunner {
+
+    private static final String LOG_TAG = "SpoonClearingAndroidTestRunner";
+    private IShellEnabledDevice mRemoteDevice;
+    private CollectingOutputReceiver mOutputReceiver;
+    private long mMaxTimeoutMs = 10000L;
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java]: `mMaxTimeoutMs` and `mMaxTimeToOutputResponseMs` should both just be static constants

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    private long mMaxTimeToOutputResponseMs = 10000L;
+    private String mAppPackageName;
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java]: unused

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            throws TimeoutException, AdbCommandRejectedException,
+            ShellCommandUnresponsiveException, IOException {
+
+        try {
+            if (mClearAppDataBeforeEachTest) {
+                Log.w(LOG_TAG, String.format("Running adb clear command: %s", mClearCommandStr));
+                mRemoteDevice.executeShellCommand(mClearCommandStr, mOutputReceiver, mMaxTimeoutMs,
+                        mMaxTimeToOutputResponseMs, TimeUnit.MILLISECONDS);
+            }
+        } catch (IOException e) {
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java]: convert this to a multi-catch and inline the `logErrorString` method?

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java (hunk 3 above)
Comment (JakeWharton) [reply spoon-runner/src/main/java/com/squareup/spoon/SpoonClearingAndroidTestRunner.java]: Also delete all the 'm' prefixes.

**Comment by rhomel** [comment]
Thanks for the feedback. All requested changes implemented. 

FYI I changed logging to Spoon's logger instead of the Android logger. When the clear command doesn't execute as expected (doesn't output "Success") a Spoon runner log message will be displayed.

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonAndroidTestRunner.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+import static com.squareup.spoon.SpoonLogger.logDebug;
+import static com.squareup.spoon.SpoonLogger.logError;
+
+/**
+ * Allows the option of running the "pm clear" command before every "run" invocation.
+ */
+public class SpoonAndroidTestRunner extends RemoteAndroidTestRunner {
+
+  private static final long maxTimeoutMs = 10000L;
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonAndroidTestRunner.java]: these should be `MAX_TIMEOUT_MILLIS` and `MAX_TIME_TO_OUTPUT_RESPONSE_MILLIS`

**Comment by rhomel** [comment]
Feedback implemented!

**Comment by edenman** [comment]
Thanks @rhomel!

**Comment by rhomel** [comment]
Android Test Orchestrator was supposed to have this function but Google
didn't actually implement it.

I just check and it was just recently added in 1.0.2-beta1. Maybe in the
next SDK release we will have "official" access to it.

https://issuetracker.google.com/issues/77549481

**Comment by rhomel** [comment]
Actually it looks like 1.0.2 is actually available, there's just no release
notes yet...

**Pull request merged**